src/dlgPackageExporter.cpp

```diff
@@ -662,228 +665,244 @@ void dlgPackageExporter::slot_export_package()
+                    true);
+        });
+        watcher->setFuture(future);
+    }
 
-        } else {
-            // Opened/created archive file successfully
-/*
- * Previous code here failed if the user included sub-directories in the
- * temporary directory where the contents for the package/module was assembled -
- * as sub-directories are now correctly handled by the installer code we have
- * revised the code here to also work should the user use them. NB This is
- * typically the case where they have collections of images or sound files
- * which they wish to store in a heirarchical manner...! - Slysven
- */
+    if (isOk) {
+        displayResultMessage(tr("Exporting package..."), true);
+    } else {
+        // Failed - convert cancel to a close button
+        ui->buttonBox->removeButton(mCancelButton);
+        ui->buttonBox->addButton(QDialogButtonBox::Close);
+        connect(ui->buttonBox->button(QDialogButtonBox::Close), &QAbstractButton::clicked, this, &dlgPackageExporter::close);
+    }
+}
+
+std::pair<bool, QString> dlgPackageExporter::zipPackage(const QString& stagingDirName, const QString& packagePathFileName, const QString& xmlPathFileName, const QString& packageName, const QString& packageConfig)
+{
+    bool isOk = true;
+    QString error;
+    // zip error code:
+    int ze = 0;
+
+
+    // ZIP_CREATE creates the archive if it does not exist.
+    // ZIP_TRUNCATE zaps any contents in a previously existing file.
+    zip* archive = zip_open(packagePathFileName.toUtf8().constData(), ZIP_CREATE|ZIP_TRUNCATE, &ze);
+
+    if (!archive) {
+        // Failed to open/create archive file
+        // We now use the better error handling system (not requiring a
+        // previously defined finite-sized char type buffer {which obviously
+        // could have string buffer overflow issues} which is available in
+        // post 0.10 versions of libzip):
+        zip_error_t error;
+        zip_error_init_with_code(&error, ze);
+        return {false, tr("Failed to open package file. Error is: \"%1\".",
+                                // Intentional comment to separate arguments
+                                "This error message is shown when the libzip library code is unable to open the file that was to be the end result of the export process. As this may be an existing file anywhere in the computer's file-system(s) it is possible that permissions on the directory or an existing file that is to be overwritten may be a source of problems here.")
+                             .arg(zip_error_strerror(&error))};
+        zip_error_fini(&error);
+        isOk = false;
+        // The above flag will now cause execution to drop down to the bottom of
+        // the method and set up a means to close the dialogue after the user
+        // has seen the error message...
+
+    }
+    
+    // Opened/created archive file successfully
 #if defined(Q_OS_WIN32)
 /*
- * From Qt Docs:
```

Comment:
Removing the pre-`*` space messes up the block comment layout (I think Qt Creator can add the ` *` automagically - which does make block comments look a bit more visually attractive/coherent IMHO)  - can we avoid it being done?